server/reload.go

```diff
@@ -0,0 +1,99 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// FlagSnapshot captures the server options as specified by CLI flags at
+// startup. This should not be modified once the server has started.
+var FlagSnapshot = &Options{}
+
+// option is a hot-swappable configuration setting.
+type option interface {
+	// Apply the server option.
+	Apply(server *Server)
+}
+
+// traceOption implements the option interface for the `trace` setting.
+type traceOption struct {
+	newValue bool
+}
+
+// Apply the tracing change by reconfiguring the server's logger.
+func (t *traceOption) Apply(server *Server) {
+	Noticef("Reloaded: trace = %v", t.newValue)
+	server.ConfigureLogger()
+}
+
+// Reload reads the current configuration file and applies any supported
+// changes. This returns an error if the server was not started with a config
+// file or an option which doesn't support hot-swapping was changed.
+func (s *Server) Reload() error {
+	if s.configFile == "" {
+		return errors.New("Can only reload config when a file is provided using -c or --config")
+	}
+	newOpts, err := ProcessConfigFile(s.configFile)
+	if err != nil {
+		// TODO: Dump previous good config to a .bak file?
+		return fmt.Errorf("Config reload failed: %s", err)
+	}
+	// Apply flags over config file settings.
+	newOpts = MergeOptions(newOpts, FlagSnapshot)
+	processOptions(newOpts)
+	return s.reloadOptions(newOpts)
+}
+
+// reloadOptions reloads the server config with the provided options. If an
+// option that doesn't support hot-swapping is changed, this returns an error.
+func (s *Server) reloadOptions(newOpts *Options) error {
+	changed, err := s.diffOptions(newOpts)
+	if err != nil {
+		return err
+	}
+	s.setOpts(newOpts)
+	s.applyOptions(changed)
+	return nil
```

Comment:
I see, but if we apply the new options to the server's options that a caller has a pointer to (getOpts()), there is a risk that the caller accesses values that are being changed on the fly.